### DISCUSSION TITLE
How to authenticate/create a session, after a user opens a magic link email? (SvelteKit)

### DISCUSSION THREAD

**Discussion by sbutler-gh** [supabase/supabase#1977] [Questions]
I'm working with SvelteKit and Supabase, and here is my situation:

1. A new user can sign-up for the service by just entering their email address, which sends them a magic link
2. When the new user clicks the magic link in their email, it redirects them to the site, with an access token in the URL

**How do I create a new session for the new user, after they click the magic link and they're redirected to the page?**

I suppose I'll have to do something with the access token?

On my `__layout.svelte` component, I'm currently pulling data from supabase on page load like this (via `/loadprojects` endpoint):

```
export const load = async ({ fetch }) => {
  const res = await fetch('/loadprojects',{
    method: 'get'
  });

  if (res.ok) {
    const todos = await res.json();

    return {
      props: { 
        projects: todos.table
      }

    };
  }

  const { message } = await res.json();

  return {
    error: new Error(message)
  };
};
```

Would I add something related to the access token and session in that function?

Or somewhere in `hooks.js`?  Currently, that just looks like this:

```
import cookie from 'cookie'

/** @type {import('@sveltejs/kit').GetSession} */
export function getSession(request) {
	return cookie.parse(request.headers.cookie || '').session || null
}
```

Any advice would be much appreciated.

(For context, I'm trying to do all calls to Supabase via endpoints (e.g. signin.js) for security reasons, so avoiding supabase functions in Svelte components)

**Comment by ashleyconnor** [comment]
I have a repo up [here](https://github.com/ashleyconnor/sveltekit-supabase-demo), it still needs work but hopefully it demonstrates how to get your auth sessions working with SvelteKit.

Post-redirect Supabase will fire the callback you set via `supabase.auth.onAuthStateChange` which I have in my `__layout.svelte`.

```javascript
import { supabase } from "$lib/supabaseClient";
import { session } from '$app/stores';
import { setAuthCookie } from '$lib/utils/session';

supabase.auth.onAuthStateChange(async (event, _session) => {
  if (event !== "SIGNED_OUT") {
    session.set({user: _session.user});
    await setAuthCookie(_session);
  }
});
```

`setAuthCookie` is a function that sends the token to an "auth" endpoint which responds with a cookie. This cookie allows me to make authenticated calls using the Supabase client in my endpoints thus enabling server side rendering of authenticated pages. See the `hooks.js` file for more details.

Thanks to [@one-aalam's repo](https://github.com/one-aalam/svelte-starter-kit/tree/auth-supabase) for providing a way to integrate `supabase-js` with SvelteKit's hooks.
